if_constexpr for C++14

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31092 Boxing uses if_constexpr instead of SFINAE
* **#31091 if_constexpr for C++14**

This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

smessmer committed:
if_constexpr for C++14

This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

[ghstack-poisoned]

ezyang commented:
Where did this implementation come from

smessmer committed:
Update on "if_constexpr for C++14"


This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

[ghstack-poisoned]

smessmer committed:
Update on "if_constexpr for C++14"


This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 200eab7:
* **1/7** broken upstream at merge base 409151e (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=409151e1bb90c66546901d372f59930f72242ee9&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **6/7** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=200eab7c05b0ec9314f6b3f596a434ec7ab205d2).

### 6 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962338) caffe2_onnx_py2_gcc5_ubuntu16_04_build (1/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74969140))
<details>
<summary>
<code>Dec 16 10:18:17 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:18:17                                                                  ^ 
Dec 16 10:18:17 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:18:17 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:18:17 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:18:17      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:18:17                                                                        ^ 
Dec 16 10:18:17 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:18:17 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:18:17      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:18:17                                                                  ^ 
Dec 16 10:18:17 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:18:17 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:18:17 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:18:17 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:18:17 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:18:17 CMakeFiles/Makefile2:2595: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:18:18 [ 41%] Generating src/x86_64-fma/blas/s8gemm.py.o 
Dec 16 10:18:18 [ 41%] Generating src/x86_64-fma/blas/c8gemm.py.o 
Dec 16 10:18:19 [ 41%] Generating src/x86_64-fma/blas/s4c6gemm.py.o 
Dec 16 10:18:20 [ 41%] Generating src/x86_64-fma/blas/conv1x1.py.o 
Dec 16 10:18:20 [ 41%] Generating src/x86_64-fma/blas/sgemm.py.o 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962331) pytorch_linux_xenial_py3_6_gcc5_4_build (2/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74970347))
<details>
<summary>
<code>Dec 16 10:19:45 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:19:45                                                                  ^ 
Dec 16 10:19:45 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:19:45 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:19:45 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:19:45      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:19:45                                                                        ^ 
Dec 16 10:19:45 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:19:45 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:19:45      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:19:45                                                                  ^ 
Dec 16 10:19:45 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:19:45 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:19:45 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:19:45 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:19:45 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:19:45 CMakeFiles/Makefile2:2396: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:19:47 [ 42%] Linking CXX executable ../../bin/c10_typeid_test 
Dec 16 10:19:47 [ 42%] Built target c10_typeid_test 
Dec 16 10:19:53 [ 42%] Generating src/x86_64-fma/2d-winograd-8x8-3x3.py.o 
Dec 16 10:19:55 [ 42%] Generating src/x86_64-fma/blas/s8gemm.py.o 
Dec 16 10:19:55 [ 42%] Generating src/x86_64-fma/blas/c8gemm.py.o 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962347) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (3/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74970348))
<details>
<summary>
<code>Dec 16 10:18:53 FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o</code>
</summary>

```
Dec 16 10:18:52 [753/2999] Linking CXX executable bin/c10_tempfile_test 
Dec 16 10:18:52 [754/2999] Building CXX object c10/test/CMakeFiles/c10_ConstexprCrc_test.dir/util/ConstexprCrc_test.cpp.o 
Dec 16 10:18:52 [755/2999] Linking CXX executable bin/c10_ConstexprCrc_test 
Dec 16 10:18:52 [756/2999] Building CXX object c10/test/CMakeFiles/c10_TensorTypeSet_test.dir/core/TensorTypeSet_test.cpp.o 
Dec 16 10:18:52 [757/2999] Building CXX object c10/test/CMakeFiles/c10_string_view_test.dir/util/string_view_test.cpp.o 
Dec 16 10:18:52 [758/2999] Linking CXX executable bin/c10_TensorTypeSet_test 
Dec 16 10:18:52 [759/2999] Linking CXX executable bin/c10_string_view_test 
Dec 16 10:18:53 [760/2999] Building CXX object c10/test/CMakeFiles/c10_TypeTraits_test.dir/util/TypeTraits_test.cpp.o 
Dec 16 10:18:53 [761/2999] Linking CXX executable bin/c10_TypeTraits_test 
Dec 16 10:18:53 [762/2999] Building CXX object c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o 
Dec 16 10:18:53 FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o  
ror=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -O3  -fPIE   -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std=gnu++14 -MD -MT c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o -MF c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o.d -o c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o -c ../c10/test/util/C++17_test.cpp 
Dec 16 10:18:53 ../c10/test/util/C++17_test.cpp: In member function ‘virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withIdentityArg_Test::TestBody()’: 
Dec 16 10:18:53 ../c10/test/util/C++17_test.cpp:99:71: error: expected ‘,’ before ‘)’ token 
Dec 16 10:18:53      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:18:53                                                                        ^ 
Dec 16 10:18:53 ../c10/test/util/C++17_test.cpp:99:71: error: expected string-literal before ‘)’ token 
Dec 16 10:18:53 ../c10/test/util/C++17_test.cpp:105:65: error: expected ‘,’ before ‘)’ token 
Dec 16 10:18:53      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:18:53                                                                  ^ 
Dec 16 10:18:53 ../c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ‘)’ token 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962346) pytorch_windows_build (4/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74974778))
<details>
<summary>
<code>FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.obj</code>
</summary>

```
 
423~1.281\bin\Hostx64\x64\link.exe  c10\test\CMakeFiles\c10_TypeIndex_test.dir\util\TypeIndex_test.cpp.obj  /out:bin\c10_TypeIndex_test.exe /implib:LIB\c10_TypeIndex_test.lib /pdb:bin\c10_TypeIndex_test.pdb /version:0.0  /machine:x64 /ignore:4049 /ignore:4217 /INCREMENTAL:NO /subsystem:console  LIB\c10.lib  LIB\gmock.lib  LIB\gtest.lib  LIB\gtest_main.lib  LIB\gtest.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ." 
Microsoft (R) Incremental Linker Version 14.23.28107.0 
Copyright (C) Microsoft Corporation.  All rights reserved. 
 
.exe  c10\test\CMakeFiles\c10_Metaprogramming_test.dir\util\Metaprogramming_test.cpp.obj  /out:bin\c10_Metaprogramming_test.exe /implib:LIB\c10_Metaprogramming_test.lib /pdb:bin\c10_Metaprogramming_test.pdb /version:0.0  /machine:x64 /ignore:4049 /ignore:4217 /INCREMENTAL:NO /subsystem:console  LIB\c10.lib  LIB\gmock.lib  LIB\gtest.lib  LIB\gtest_main.lib  LIB\gtest.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ." 
Microsoft (R) Incremental Linker Version 14.23.28107.0 
Copyright (C) Microsoft Corporation.  All rights reserved. 
 
etest\googlemock\include -I..\third_party\googletest\googlemock -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp:experimental /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std:c++14 /showIncludes /Foc10\test\CMakeFiles\c10_C++17_test.dir\util\C++17_test.cpp.obj /Fdc10\test\CMakeFiles\c10_C++17_test.dir\ /FS -c ..\c10\test\util\C++17_test.cpp 
FAILED: c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.obj  
etest\googlemock\include -I..\third_party\googletest\googlemock -I..\third_party\googletest\googletest\include -I..\third_party\googletest\googletest /DWIN32 /D_WINDOWS /GR  /w /EHa /MP /bigobj -openmp:experimental /MD /O2 /Ob2  /w /EHa /MP /bigobj   -DCUDA_HAS_FP16=1 -DUSE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -std:c++14 /showIncludes /Foc10\test\CMakeFiles\c10_C++17_test.dir\util\C++17_test.cpp.obj /Fdc10\test\CMakeFiles\c10_C++17_test.dir\ /FS -c ..\c10\test\util\C++17_test.cpp 
..\c10\test\util\C++17_test.cpp(99): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
..\c10\test\util\C++17_test.cpp(105): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
..\c10\test\util\C++17_test.cpp(113): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
..\c10\test\util\C++17_test.cpp(119): error C2429: language feature 'terse static assert' requires compiler flag '/std:c++17'
Microsoft (R) C/C++ Optimizing Compiler Version 19.23.28107 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

.281\bin\Hostx64\x64\link.exe  c10\test\CMakeFiles\c10_TypeTraits_test.dir\util\TypeTraits_test.cpp.obj  /out:bin\c10_TypeTraits_test.exe /implib:LIB\c10_TypeTraits_test.lib /pdb:bin\c10_TypeTraits_test.pdb /version:0.0  /machine:x64 /ignore:4049 /ignore:4217 /INCREMENTAL:NO /subsystem:console  LIB\c10.lib  LIB\gmock.lib  LIB\gtest.lib  LIB\gtest_main.lib  LIB\gtest.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ." 
Microsoft (R) Incremental Linker Version 14.23.28107.0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962330) pytorch_linux_xenial_cuda9_cudnn7_py3_build (5/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74976564))
<details>
<summary>
<code>Dec 16 10:23:02 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:23:02                                                                  ^ 
Dec 16 10:23:02 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:23:02 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:23:02 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:23:02      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:23:02                                                                        ^ 
Dec 16 10:23:02 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:23:02 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:23:02      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:23:02                                                                  ^ 
Dec 16 10:23:02 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:23:02 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:23:02 CMakeFiles/Makefile2:2736: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:23:02 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:23:02 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:23:02 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:23:05 Compiling  all_reduce.cu                       > /var/lib/jenkins/workspace/build/nccl/obj/collectives/device/all_reduce_prod_f64.o 
Dec 16 10:23:05 Scanning dependencies of target nnpack 
Dec 16 10:23:05 [ 35%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/init.c.o 
Dec 16 10:23:05 [ 35%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/convolution-inference.c.o 
Dec 16 10:23:05 [ 35%] Building C object confu-deps/NNPACK/CMakeFiles/nnpack.dir/src/fully-connected-inference.c.o 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3962349) pytorch_libtorch_linux_xenial_cuda9_cudnn7_py3_build (6/6)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74978762))
<details>
<summary>
<code>Dec 16 10:22:30 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token</code>
</summary>

```
Dec 16 10:22:30                                                                  ^ 
Dec 16 10:22:30 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:105:65: error: expected string-literal before ')' token 
Dec 16 10:22:30 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp: In member function 'virtual void {anonymous}::test_if_constexpr::if_constexpr_returnTypeCanDiffer_withoutIdentityArg_Test::TestBody()': 
Dec 16 10:22:30 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected ',' before ')' token 
Dec 16 10:22:30      static_assert(std::is_same<std::string, decltype(a_string)>::value); 
Dec 16 10:22:30                                                                        ^ 
Dec 16 10:22:30 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:113:71: error: expected string-literal before ')' token 
Dec 16 10:22:30 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected ',' before ')' token 
Dec 16 10:22:30      static_assert(std::is_same<int64_t, decltype(an_int)>::value); 
Dec 16 10:22:30                                                                  ^ 
Dec 16 10:22:30 /var/lib/jenkins/workspace/c10/test/util/C++17_test.cpp:119:65: error: expected string-literal before ')' token 
Dec 16 10:22:30 c10/test/CMakeFiles/c10_C++17_test.dir/build.make:62: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o' failed 
Dec 16 10:22:30 make[2]: Leaving directory '/var/lib/jenkins/cpp-build/caffe2/build' 
Dec 16 10:22:30 make[2]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/util/C++17_test.cpp.o] Error 1 
Dec 16 10:22:30 make[1]: *** [c10/test/CMakeFiles/c10_C++17_test.dir/all] Error 2 
Dec 16 10:22:30 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 10:22:30 CMakeFiles/Makefile2:2497: recipe for target 'c10/test/CMakeFiles/c10_C++17_test.dir/all' failed 
Dec 16 10:22:30 [ 27%] Generating src/x86_64-fma/2d-winograd-8x8-3x3.py.o 
_party/python-six:/var/lib/jenkins/workspace/third_party/python-peachpy /opt/conda/bin/python -m peachpy.x86_64 -mabi=sysv -g4 -mimage-format=elf -I/var/lib/jenkins/workspace/third_party/NNPACK/src -I/var/lib/jenkins/workspace/third_party/NNPACK/src/x86_64-fma -I/var/lib/jenkins/workspace/third_party/FP16/include -o /var/lib/jenkins/cpp-build/caffe2/build/confu-deps/NNPACK/src/x86_64-fma/2d-winograd-8x8-3x3.py.o /var/lib/jenkins/workspace/third_party/NNPACK/src/x86_64-fma/2d-winograd-8x8-3x3.py 
Dec 16 10:22:32 nvcc warning : '--device-debug (-G)' overrides '--generate-line-info (-lineinfo)' 
Dec 16 10:22:32 ptxas warning : Conflicting options --device-debug and --generate-line-info specified, ignoring --generate-line-info option 
```

</details>

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [caffe2_onnx_py3_6_clang7_ubuntu16_04_test](https://dr.pytorch.org/build-details.html?build_id=75046685)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

smessmer committed:
Update on "if_constexpr for C++14"


This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

[ghstack-poisoned]

smessmer committed:
Update on "if_constexpr for C++14"


This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

[ghstack-poisoned]

smessmer commented:
> Where did this implementation come from

I wrote it a couple years back for a different project

smessmer committed:
Update on "if_constexpr for C++14"


This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

[ghstack-poisoned]

ezyang commented on c10/util/C++17.h:
```diff
@@ -154,6 +154,110 @@ invoke(Functor&& f, Args&&... args) {
+ * Note: The _ argument passed in Example 3 is the identity function, i.e. it does nothing.
+ *       It is used to force the compiler to delay type checking, because the compiler
+ *       doesn't know what kind of _ is passed in. Without it, the compiler would fail
+ *       when you try to access t.value but the member doesn't exist.
```
Although it would be dead code, it would be nice to see the reference implementation of this template which would be implementable with C++17 constexpr-if directly.

smessmer committed:
Update on "if_constexpr for C++14"


This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

[ghstack-poisoned]

smessmer committed:
Update on "if_constexpr for C++14"


This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

[ghstack-poisoned]

smessmer committed:
Update on "if_constexpr for C++14"


This implements a C++17 "if constexpr" like feature for C++14.
This can be used, for example, to replace SFINAE or to force the compiler to remove some parts of a function in the assembly based on a condition.
PRs stacked on top will use this to simplify some of our template metaprogramming.

Differential Revision: [D18927220](https://our.internmc.facebook.com/intern/diff/D18927220/)

[ghstack-poisoned]

mruberry commented:
Unlanding. This appears to have broken ci/circleci: pytorch_windows_vs2017_14.11_py36_cpu_build and ci/circleci: pytorch_windows_vs2017_14.11_py36_cuda10.1_build. 31092 was already unlanded.
